Disable building of shared library needed by collect_stats.py by default

shunghsiyu opened the pull request:
The `_phoebe.abi3.so` shared library is used by collect_stats.py to retrieve various system metric; and the shared library itself requires python3-cffi to generate the binding (`_phoebe.c`).

On Linux distribution that doesn't come with a python3-cffi package is becomes a problem, while there are other ways to install python3-cffi (e.g. `pip3 install cffi`), doing so it not ideal when building the phoebe package itself.

Thus the easiest way to deal with the problem is disable building of `_phoebe.abi3.so` by default so packaging doesn't need python3-cffi, and explicitly enable it in our CI.

shunghsiyu committed:
Introduce 'collector' feature

So that users can enforce the module needed by collect_stats.py being present,
and fail the configure step if otherwise.

shunghsiyu committed:
Enable the collector feature in CI

shunghsiyu commented:
@dcermak since you know meson and GitHub action much better, can you take a quick glance?

dcermak reviewed (changes requested):
This should be done via a meson feature, or even better, the build of the shared library should depend on the existence of the ffi module

shunghsiyu commented:
> build of the shared library should depend on the existence of the ffi module

Interesting, haven't thought of that, let me try to do so instead.

dcermak commented:
I can send a PR instead, if you don't like to mess with this.

shunghsiyu commented:
I'm actually fine with this; finally got a chance to learn meson with someone who knows it ;)

dcermak commented:
> I'm actually fine with this; finally got a chance to learn meson with someone who knows it ;)

you'll need this: https://mesonbuild.com/Python-module.html#find_installation and https://mesonbuild.com/Build-options.html#features

shunghsiyu committed:
Detect the presence of python3-cffi module

Use the 'module' argument of find_installation() to check whether Python's CFFI
module is installed. If it is then build the shared library that is needed by
collect_stats.py, which is a script used to collect system metrics for AI model
training; if a Python installation with CFFI module is not found, then don't
bother building the shared library. (Note: collect_stats.py is generally not
needed on production/deployment)

A major behind this is that there are SLES distributions that does not provide
the python3-cffi package, which makes RPM packaging difficult. With building of
the shared library (requires python3-cffi) determined dynmically, we can get
away with the missing python3-cffi package.

shunghsiyu commented:
Now use `find_installation('python3', required : false, modules : ['cffi'])` instead to find a Python installation with CFFI module available.

I was hoping to get rid of the `if` conditional and replace it with `required` inside `custom_target()` and `shared_library()`. Sadly `custom_target()` doesn't seem to support the `required` argument.

dcermak commented on scripts/meson.build:
```diff
@@ -1,14 +1,16 @@
-prog_python = import('python').find_installation('python3')
+prog_python = import('python').find_installation('python3', required : false, modules : ['cffi'])
```
You could try to use the `required` value from a [feature](https://mesonbuild.com/Build-options.html#features), so that users can enforce this module being present (and fail the configure step).

shunghsiyu replied:
Done, thanks for the tips!

mvarlese commented:
Thanks!
